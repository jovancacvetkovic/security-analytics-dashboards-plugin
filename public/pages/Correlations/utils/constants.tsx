/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { CorrelationGraphData } from '../../../../types';
import { ruleSeverity, ruleTypes } from '../../Rules/utils/constants';
import { FilterItem } from '../components/FilterGroup';
import { EuiIcon } from '@elastic/eui';

export const graphRenderOptions = {
  nodes: {
    shape: 'circle',
  },
  edges: {
    arrows: {
      to: {
        enabled: false,
      },
    },
    color: '#000000',
    chosen: true,
  },
  layout: {
    hierarchical: false,
    randomSeed: 2222,
    improvedLayout: false,
  },
  autoResize: true,
  height: '800px',
  width: '100%',
  physics: {
    stabilization: {
      fit: true,
      iterations: 1000,
    },
  },
  interaction: {
    zoomView: true,
    zoomSpeed: 0.2,
    dragView: true,
    dragNodes: false,
    multiselect: true,
    tooltipDelay: 50,
    hover: true,
  },
};

export const defaultLogTypeFilterItemOptions: FilterItem[] = Object.values(ruleTypes).map(
  (type) => {
    return {
      name: `${type.abbr}: ${type.label}`,
      id: type.value,
      checked: 'on',
    };
  }
);

export const defaultSeverityFilterItemOptions: FilterItem[] = Object.values(ruleSeverity).map(
  (sev) => {
    return {
      name: (
        <p>
          <EuiIcon type={'dot'} color={sev.color} /> {`${sev.priority} ${sev.name}`}
        </p>
      ),
      id: sev.value,
      checked: 'on',
    };
  }
);

export const emptyGraphData: CorrelationGraphData = {
  graph: {
    nodes: [],
    edges: [],
  },
  events: {},
};

export const getAbbrFromLogType = (logType: string) => {
  return ruleTypes.find((ruleType) => ruleType.value === logType)?.abbr || '-';
};

export const getLabelFromLogType = (logType: string) => {
  return ruleTypes.find((ruleType) => ruleType.value === logType)?.label || '-';
};

export const getSeverityLabel = (sev: string) => {
  return ruleSeverity.find((severity) => severity.value === sev)?.name || '';
};

export const getSeverityColor = (sev: string) => {
  return ruleSeverity.find((severity) => severity.value === sev)?.color || 'black';
};
