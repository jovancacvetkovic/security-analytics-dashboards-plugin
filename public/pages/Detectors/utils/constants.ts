/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const MAX_DETECTOR_INPUTS = 1;
export const MIN_NUM_DATA_SOURCES = 1;

export const STATUS_OPTIONS = [
  { value: 'test_status1', text: 'Test status #1' },
  { value: 'test_status2', text: 'Test status #2' },
];

export const EMPTY_DEFAULT_PERIOD_SCHEDULE = {
  period: {
    interval: 1,
    unit: 'm',
  },
};

export const EMPTY_DEFAULT_DETECTOR_INPUT = {
  input: {
    description: '',
    indices: [],
    rules: [],
  },
};

export const DETECTOR_TYPES = {
  NETWORK: { id: 'network', label: 'Network events', abbr: 'NTW' },
  DNS: { id: 'dns', label: 'DNS logs', abbr: 'DNS' },
  APACHE_ACCESS: { id: 'apache_access', label: 'Apache access logs', abbr: 'APC' },
  WINDOWS: { id: 'windows', label: 'Windows logs', abbr: 'WIN' },
  AD_LDAP: { id: 'ad_ldap', label: 'AD/LDAP logs', abbr: 'AD' },
  SYSTEM: { id: 'linux', label: 'System logs', abbr: 'LNX' },
  CLOUD_TRAIL: { id: 'cloudtrail', label: 'Cloud Trail logs', abbr: 'CLT' },
  S3: { id: 's3', label: 'S3 access logs', abbr: 'S3' },
  GWORKSPACE: { id: 'gworkspace', label: 'Google Workspace logs', abbr: 'GGL' },
  GITHUB: { id: 'github', label: 'Github actions', abbr: 'GHB' },
  M365: { id: 'm365', label: 'Microsoft 365 logs', abbr: 'MSO' },
  OKTA: { id: 'okta', label: 'Okta events', abbr: 'OKT' },
  AZURE: { id: 'azure', label: 'Azure logs', abbr: 'AZR' },
};
