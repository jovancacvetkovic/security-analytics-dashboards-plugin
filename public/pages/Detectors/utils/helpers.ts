/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { DetectorHit } from '../../../../server/models/interfaces';
import { FieldMappingsTableItem } from '../../../../types';

export function getDetectorIds(detectors: DetectorHit[]) {
  return detectors.map((detector) => detector._id).join(', ');
}

export function getDetectorNames(detectors: DetectorHit[]) {
  return detectors.map((detector) => detector._source.name).join(', ');
}

export const getMappingFields = (
  properties: any,
  items: FieldMappingsTableItem[] = [],
  prefix: string = ''
): FieldMappingsTableItem[] => {
  for (let field in properties) {
    const fullFieldName = prefix ? `${prefix}.${field}` : field;
    const nextProperties = properties[field].properties;
    if (!nextProperties) {
      items.push({
        ruleFieldName: fullFieldName,
        logFieldName: properties[field].path,
      });
    } else {
      getMappingFields(nextProperties, items, fullFieldName);
    }
  }
  return items;
};
