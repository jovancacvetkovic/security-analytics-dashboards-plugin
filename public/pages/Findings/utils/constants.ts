/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export enum FindingFlyoutTabId {
  DETAILS = 'Details',
  CORRELATIONS = 'Correlations',
}

export const FindingFlyoutTabs = [
  { id: FindingFlyoutTabId.DETAILS, name: 'Details' },
  { id: FindingFlyoutTabId.CORRELATIONS, name: 'Correlations' },
];
