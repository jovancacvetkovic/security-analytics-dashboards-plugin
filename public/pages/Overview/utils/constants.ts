/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const summaryGroupByOptions = [
  { text: 'All findings', value: 'finding' },
  { text: 'Log type', value: 'logType' },
];
