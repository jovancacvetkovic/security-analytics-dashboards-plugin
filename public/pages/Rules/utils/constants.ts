/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { euiPaletteForStatus } from '@elastic/eui';

export const ruleTypes: { label: string; value: string; abbr: string }[] = [
  { abbr: 'NTW', label: 'Network', value: 'network' },
  { abbr: 'DNS', label: 'DNS', value: 'dns' },
  { abbr: 'APC', label: 'Apache Access', value: 'apache_access' },
  { abbr: 'WIN', label: 'Windows', value: 'windows' },
  { abbr: 'AD', label: 'AD/LDAP', value: 'ad_ldap' },
  { abbr: 'LNX', label: 'Linux', value: 'linux' },
  { abbr: 'CLT', label: 'Cloudtrail', value: 'cloudtrail' },
  { abbr: 'S3', label: 'S3', value: 's3' },
  { abbr: 'GGL', label: 'Google Workspace', value: 'gworkspace' },
  { abbr: 'GHB', label: 'Github actions', value: 'github' },
  { abbr: 'MSO', label: 'Microsoft 365', value: 'm365' },
  { abbr: 'OKT', label: 'Okta', value: 'okta' },
  { abbr: 'AZR', label: 'Azure', value: 'azure' },
];

const paletteColors = euiPaletteForStatus(5);

export const ruleSeverity: { name: string; value: string; priority: string; color: string }[] = [
  { name: 'Critical', value: 'critical', priority: '1', color: paletteColors[4] },
  { name: 'High', value: 'high', priority: '2', color: paletteColors[3] },
  { name: 'Medium', value: 'medium', priority: '3', color: paletteColors[2] },
  { name: 'Low', value: 'low', priority: '4', color: paletteColors[1] },
  { name: 'Informational', value: 'informational', priority: '5', color: paletteColors[0] },
];

export const ruleSource: string[] = ['Sigma', 'Custom'];

export const ruleStatus: string[] = ['experimental', 'test', 'stable'];
