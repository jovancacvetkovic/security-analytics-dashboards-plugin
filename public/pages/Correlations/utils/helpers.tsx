/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { EuiBasicTableColumn, EuiBadge, EuiToolTip, EuiButtonIcon, EuiLink } from '@elastic/eui';
import {
  ArgsWithError,
  ArgsWithQuery,
  CorrelationRule,
  CorrelationRuleQuery,
} from '../../../../types';
import { Search } from '@opensearch-project/oui/src/eui_components/basic_table';
import { ruleTypes } from '../../Rules/utils/constants';
import { FieldClause } from '@opensearch-project/oui/src/eui_components/search_bar/query/ast';

export const getCorrelationRulesTableColumns = (
  onRuleNameClick: (rule: CorrelationRule) => void,
  _refreshRules: (ruleItem: CorrelationRule) => void
): EuiBasicTableColumn<CorrelationRule>[] => {
  return [
    {
      field: 'name',
      name: 'Name',
      sortable: true,
      truncateText: true,
      render: (name: string, ruleItem: CorrelationRule) => (
        <EuiLink onClick={() => onRuleNameClick(ruleItem)}>{name}</EuiLink>
      ),
    },
    {
      name: 'Log types',
      render: (ruleItem: CorrelationRule) => {
        const badges = [...new Set(ruleItem.queries?.map((query) => query.logType))];
        return (
          <>
            {badges.map((badge) => (
              <EuiBadge color="hollow">{badge}</EuiBadge>
            ))}
          </>
        );
      },
    },
    {
      field: 'queries',
      name: 'Queries',
      align: 'right',
      render: (queries: CorrelationRuleQuery[], ruleItem: CorrelationRule) => {
        return queries.length;
      },
      width: '10%',
    },
    {
      name: 'Actions',
      field: '',
      actions: [
        {
          render: (ruleItem: CorrelationRule) => (
            <EuiToolTip content={'Delete'}>
              <EuiButtonIcon
                aria-label={'Delete correlation rule'}
                data-test-subj={`view-details-icon`}
                iconType={'trash'}
                color="danger"
                onClick={() => _refreshRules(ruleItem)}
              />
            </EuiToolTip>
          ),
        },
      ],
    },
  ];
};

export const getCorrelationRulesTableSearchConfig = (
  onLogTypeFilterChange: (logTypes?: string[]) => void
): Search => {
  return {
    box: {
      placeholder: 'Search by rule name, log type?',
    },
    onChange: (args: ArgsWithQuery | ArgsWithError) => {
      if (!args.error) {
        const logTypeFieldClauseIdx = args.query.ast.clauses.findIndex(
          (clause) => clause.type === 'field' && clause.field === 'logTypes'
        );
        const logTypeFieldClause =
          logTypeFieldClauseIdx > -1 ? args.query.ast.clauses[logTypeFieldClauseIdx] : undefined;

        if (logTypeFieldClause) {
          const logTypes = (logTypeFieldClause as FieldClause).value as string[];
          // Need to remove the logTypes clause so that in built search doesn't try to apply it because that will return 0 results,
          // since it requires custom logic we implemented in the `onLogTypeFilterChange` callback
          args.query.ast.removeOrFieldClauses('logTypes');
          onLogTypeFilterChange(logTypes);
        } else if (args.query.ast.clauses.length === 0) {
          onLogTypeFilterChange(undefined);
        }
      }

      return true;
    },
    filters: [
      {
        type: 'field_value_selection',
        field: 'logTypes',
        name: 'Log Types',
        multiSelect: 'or',
        options: ruleTypes.map(({ value, label }) => ({
          value,
          name: label,
        })),
      },
    ],
  };
};
