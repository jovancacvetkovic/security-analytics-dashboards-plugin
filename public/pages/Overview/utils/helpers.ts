/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import moment from 'moment';
import { euiPaletteColorBlind, euiPaletteForStatus } from '@elastic/eui';
import { TopLevelSpec } from 'vega-lite';
import { SummaryData } from '../components/Widgets/Summary';
import dateMath from '@elastic/datemath';
import { TimeUnitsMap } from './constants';
import _ from 'lodash';
import { DEFAULT_DATE_RANGE } from '../../../utils/constants';

export type DateOpts = {
  timeUnit: string;
  dateFormat: string;
  domain?: number[];
};

/**
 * Legend selection config for the chart layer
 */
const legendSelectionCfg = {
  selection: {
    series: {
      type: 'multi',
      encodings: ['color'],
      on: 'click',
      bind: 'legend',
    },
  },
  encoding: {
    opacity: {
      condition: { selection: 'series', value: 1 },
      value: 0.2,
    },
  },
};

/**
 * Adds interactive legends to the chart layer
 * @param layer
 */
const addInteractiveLegends = (layer: any) => _.defaultsDeep(layer, legendSelectionCfg);

function getVisualizationSpec(description: string, data: any, layers: any[]): TopLevelSpec {
  return {
    config: {
      view: { stroke: 'transparent' },
      legend: {
        labelColor: '#343741',
        titleColor: '#1a1c21',
        labelFontSize: 14,
        titleFontWeight: 600,
        titleLineHeight: 21,
        titleFontSize: 14,
        titlePadding: 10,
        rowPadding: 6,
        labelFont:
          '"Inter UI", -apple-system, BlinkMacSystemFont, "Segoe UI", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"',
      },
    },
    $schema: 'https://vega.github.io/schema/vega-lite/v5.json',
    description: description,
    data: {
      values: data,
    },
    layer: layers,
  };
}

export function getOverviewVisualizationSpec(
  visualizationData: SummaryData[],
  groupBy: string,
  dateOpts: DateOpts = {
    timeUnit: 'yearmonthdatehoursminutes',
    dateFormat: '%Y-%m-%d %H:%M',
  }
): TopLevelSpec {
  const aggregate = 'sum';
  const findingsEncoding: { [x: string]: any } = {
    x: {
      timeUnit: dateOpts.timeUnit,
      field: 'time',
      title: '',
      axis: { grid: false, ticks: false, format: dateOpts.dateFormat },
    },
    y: {
      aggregate,
      field: 'finding',
      type: 'quantitative',
      title: 'Count',
      axis: { grid: true, ticks: false },
    },
    tooltip: [{ field: 'finding', aggregate: 'sum', type: 'quantitative', title: 'Findings' }],
  };

  if (groupBy === 'logType') {
    findingsEncoding['color'] = {
      field: 'logType',
      type: 'nominal',
      title: 'Log type',
      scale: {
        range: euiPaletteColorBlind(),
      },
    };
  }

  const lineColor = '#ff0000';
  return getVisualizationSpec(
    'Plot showing average data with raw values in the background.',
    visualizationData,
    [
      addInteractiveLegends({
        mark: 'bar',
        encoding: findingsEncoding,
      }),
      {
        mark: {
          type: 'line',
          color: lineColor,
          point: {
            filled: true,
            fill: lineColor,
          },
        },
        encoding: {
          x: {
            timeUnit: dateOpts.timeUnit,
            field: 'time',
            title: '',
            axis: { grid: false, ticks: false, format: dateOpts.dateFormat },
          },
          y: {
            aggregate: 'sum',
            field: 'alert',
            title: 'Count',
            axis: { grid: true, ticks: false },
          },
          tooltip: [{ field: 'alert', aggregate: 'sum', title: 'Alerts' }],
        },
      },
    ]
  );
}

export const getDomainRange = (
  range: string[] = [DEFAULT_DATE_RANGE.start, DEFAULT_DATE_RANGE.end]
): number[] => {
  return [
    dateMath.parse(range[0])?.toDate().getTime(),
    dateMath.parse(range[1])?.toDate().getTime(),
  ];
};

/**
 * Returns chart x-axis date format based on time span
 * @param start
 * @param end
 */
export function getDateFormatByTimeUnit(start: string, end: string) {
  const startMoment = dateMath.parse(start);
  const endMoment = dateMath.parse(end);
  let dateFormat = '%Y-%m-%d %H:%M';

  if (startMoment && endMoment) {
    const startData = startMoment.toObject();
    const endData = endMoment.toObject();
    const dateDiff = endMoment.diff(startMoment);
    const momentDiff = moment.duration(dateDiff);
    const daysDiff = _.get(momentDiff, '_data.days', 0);

    if (startData.years === endData.years) {
      if (startData.months !== endData.months) {
        dateFormat = '%Y-%m-%d';

        if (daysDiff < 30 && daysDiff > 1) {
          dateFormat = '%Y-%m-%d %H:%M';
        }
      } else {
        dateFormat = '%Y-%m-%d %H:%M';

        if (startData.date === endData.date) {
          dateFormat = '%H:%M:%S';
        }
      }
    }
  }

  return dateFormat;
}

export function getFindingsVisualizationSpec(
  visualizationData: any[],
  groupBy: string,
  dateOpts: DateOpts = {
    timeUnit: 'yearmonthdatehoursminutes',
    dateFormat: '%Y-%m-%d %H:%M',
    domain: [DEFAULT_DATE_RANGE.start, DEFAULT_DATE_RANGE.end],
  }
) {
  console.log('TIME UNIT', dateOpts.timeUnit);
  // dateOpts.timeUnit = 'yearmonthdatehours';
  return getVisualizationSpec('Findings data overview', visualizationData, [
    addInteractiveLegends({
      mark: {
        type: 'bar',
        strokeWidth: 1,
        stroke: 'white',
        clip: true,
      },
      encoding: {
        tooltip: [
          {
            field: 'finding',
            aggregate: 'sum',
            type: 'quantitative',
            title: 'Findings',
          },
          {
            field: 'time',
            type: 'temporal',
            title: 'Time',
            format: '%Y-%m-%d %H:%M',
          },
        ],
        x: {
          timeUnit: dateOpts.timeUnit,
          field: 'time',
          title: '',
          type: 'temporal',
          axis: {
            grid: false,
            format: dateOpts.dateFormat,
          },
          scale: {
            type: 'time',
            field: 'time',
            domain: dateOpts.domain,
          },
        },
        y: {
          aggregate: 'sum',
          field: 'finding',
          type: 'quantitative',
          title: 'Count',
          axis: { grid: true, ticks: false },
        },
        color: {
          field: groupBy,
          type: 'nominal',
          title: groupBy === 'logType' ? 'Log type' : 'Rule severity',
          scale: {
            range: euiPaletteColorBlind(),
          },
        },
      },
    }),
  ]);
}

export function getAlertsVisualizationSpec(
  visualizationData: any[],
  groupBy: string,
  dateOpts: DateOpts = {
    timeUnit: 'yearmonthdatehoursminutes',
    dateFormat: '%Y-%m-%d %H:%M',
  }
) {
  return getVisualizationSpec('Alerts data overview', visualizationData, [
    addInteractiveLegends({
      mark: 'bar',
      encoding: {
        tooltip: [{ field: 'alert', aggregate: 'sum', title: 'Alerts' }],
        x: {
          timeUnit: dateOpts.timeUnit,
          field: 'time',
          title: '',
          axis: {
            grid: false,
            ticks: false,
            format: dateOpts.dateFormat,
          },
        },
        y: {
          aggregate: 'sum',
          field: 'alert',
          type: 'quantitative',
          title: 'Count',
          axis: { grid: true, ticks: false },
        },
        color: {
          field: groupBy,
          type: 'nominal',
          title: groupBy === 'status' ? 'Alert status' : 'Alert severity',
          scale: {
            range: groupBy === 'status' ? euiPaletteForStatus(5) : euiPaletteColorBlind(),
          },
        },
      },
    }),
  ]);
}

export function getTopRulesVisualizationSpec(visualizationData: any[]) {
  return getVisualizationSpec('Most frequent detection rules', visualizationData, [
    {
      mark: { type: 'arc', innerRadius: 90 },
      transform: [
        {
          joinaggregate: [
            {
              op: 'sum',
              field: 'count',
              as: 'total',
            },
          ],
        },
        {
          calculate: 'datum.count/datum.total',
          as: 'percentage',
        },
      ],
      encoding: {
        tooltip: [
          {
            field: 'percentage',
            title: 'Percentage',
            type: 'quantitative',
            format: '2.0%',
          },
        ],
        theta: { aggregate: 'sum', field: 'count', type: 'quantitative' },
        color: {
          field: 'ruleName',
          type: 'nominal',
          title: 'Rule name',
          scale: {
            range: euiPaletteColorBlind(),
          },
        },
      },
    },
  ]);
}

export function getTimeWithMinPrecision(time: number | string) {
  const date = new Date(time);
  date.setSeconds(0);
  date.setMilliseconds(0);

  return date.getTime();
}

/**
 * Returns timeUnit based on how big time diff is between start and end dates
 * @param start Chart start time
 * @param end Chart end time
 * @param [defaultUnit = 'yearmonthdatehoursminutes'] Default timeUnit
 */
export function getChartTimeUnit(
  start: string,
  end: string,
  defaultUnit: string = 'yearmonthdatehoursminutes'
): string {
  const startMoment = dateMath.parse(start);
  const endMoment = dateMath.parse(end);
  let timeUnit: string = 'yearmonthdatehoursminutes';

  if (!startMoment || !endMoment) return defaultUnit;

  try {
    const timeDiff = endMoment.diff(startMoment);
    const momentTimeDiff = moment.duration(timeDiff);

    console.log('TIME UNIT', momentTimeDiff);
    const { years, months, days, hours, minutes, seconds } = momentTimeDiff._data;

    if (!years) {
      if (!months) {
        if (!days) {
          if (!hours) {
            if (minutes < 60) {
              timeUnit = 'yearmonthdatehoursminutes';
            }
          } else {
            if (hours < 60) {
              timeUnit = 'yearmonthdatehours';
            }
          }
        } else {
          if (days === 1) {
            timeUnit = 'yearmonthdatehours';
          } else if (days <= 14) {
            timeUnit = 'yearmonthdate';
          } else {
            timeUnit = 'yearmonthdate';
          }
        }
      } else {
        if (months <= 6) {
          timeUnit = 'yearmonthdate';
        } else {
          timeUnit = 'yearmonth';
        }
      }
    } else if (years <= 6) {
      timeUnit = 'yearmonth';

      if (years > 2) {
        timeUnit = 'yearquarter';
      }
    } else {
      timeUnit = 'year';
    }
  } catch (e) {
    console.error(`Time diff can't be calculated for dates: ${start} and ${end}`);
  }

  return timeUnit;
}
