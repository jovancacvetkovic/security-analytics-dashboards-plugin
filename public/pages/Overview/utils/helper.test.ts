import { getChartTimeUnit } from './helpers';
import { TimeUnitsMap } from './constants';

describe('helper utilities spec', () => {
  describe('tests getChartTimeUnit function', () => {
    const defaultTimeUnit = 'yearmonthdatehoursminutes';
    it(' - function should return default timeUnit if fn params are invalid', () => {
      expect(getChartTimeUnit('', '')).toBe(defaultTimeUnit);
    });

    it(' - function should return default timeUnit if one is passed as param', () => {
      const defaultFormat = 'yearmonthdate';
      expect(getChartTimeUnit('', '', defaultFormat)).toBe(defaultFormat);
    });

    const timeUnits: {
      [key: string]: string;
    } = {
      minutes: 'now-15m',
      hours: 'now-15h',
      days: 'now-15d',
      weeks: 'now-15w',
      months: 'now-5M',
      years: 'now-15y',
    };

    for (const [unit, start] of Object.entries(timeUnits)) {
      it(` - function should return ${TimeUnitsMap[unit]} if unit is ${unit}`, () => {
        expect(getChartTimeUnit(start, 'now')).toBe(TimeUnitsMap[unit]);
      });
    }
  });
});
