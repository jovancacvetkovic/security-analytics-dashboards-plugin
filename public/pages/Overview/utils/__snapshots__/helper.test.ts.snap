// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helper utilities spec tests getAlertsVisualizationSpec function  - snapshot test: should match alerts spec 1`] = `
Object {
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "config": Object {
    "legend": Object {
      "labelColor": "#343741",
      "labelFont": "\\"Inter UI\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\"",
      "labelFontSize": 14,
      "rowPadding": 6,
      "titleColor": "#1a1c21",
      "titleFontSize": 14,
      "titleFontWeight": 600,
      "titleLineHeight": 21,
      "titlePadding": 10,
    },
    "view": Object {
      "stroke": "transparent",
    },
  },
  "data": Object {
    "values": Array [],
  },
  "description": "Alerts data overview",
  "layer": Array [
    Object {
      "encoding": Object {
        "color": Object {
          "field": "",
          "scale": Object {
            "range": Array [
              "#54B399",
              "#6092C0",
              "#D36086",
              "#9170B8",
              "#CA8EAE",
              "#D6BF57",
              "#B9A888",
              "#DA8B45",
              "#AA6556",
              "#E7664C",
            ],
          },
          "title": "Alert severity",
        },
        "opacity": Object {
          "condition": Object {
            "selection": "series",
            "value": 1,
          },
          "value": 0.2,
        },
        "tooltip": Array [
          Object {
            "aggregate": "sum",
            "axis": Object {
              "grid": true,
            },
            "field": "alert",
            "title": "Alerts",
            "type": "quantitative",
          },
          Object {
            "field": "time",
            "format": "%Y-%m-%d %H:%M:%S",
            "timeUnit": Object {
              "step": 1,
              "unit": "yearmonthdatehoursminutes",
            },
            "title": "Time",
            "type": "temporal",
          },
          Object {
            "field": "",
            "title": "Alert severity",
          },
        ],
        "x": Object {
          "axis": Object {
            "format": "%Y-%m-%d %H:%M",
            "grid": false,
          },
          "field": "time",
          "scale": Object {
            "domain": undefined,
          },
          "timeUnit": Object {
            "step": 1,
            "unit": "yearmonthdatehoursminutes",
          },
          "title": "",
          "type": "temporal",
        },
        "y": Object {
          "aggregate": "sum",
          "axis": Object {
            "grid": true,
          },
          "field": "alert",
          "title": "Count",
          "type": "quantitative",
        },
      },
      "mark": Object {
        "clip": true,
        "type": "bar",
      },
      "selection": Object {
        "series": Object {
          "bind": "legend",
          "encodings": Array [
            "color",
          ],
          "on": "click",
          "type": "multi",
        },
      },
    },
  ],
}
`;

exports[`helper utilities spec tests getFindingsVisualizationSpec function  - snapshot test: should match findings spec 1`] = `
Object {
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "config": Object {
    "legend": Object {
      "labelColor": "#343741",
      "labelFont": "\\"Inter UI\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\"",
      "labelFontSize": 14,
      "rowPadding": 6,
      "titleColor": "#1a1c21",
      "titleFontSize": 14,
      "titleFontWeight": 600,
      "titleLineHeight": 21,
      "titlePadding": 10,
    },
    "view": Object {
      "stroke": "transparent",
    },
  },
  "data": Object {
    "values": Array [],
  },
  "description": "Findings data overview",
  "layer": Array [
    Object {
      "encoding": Object {
        "color": Object {
          "field": "",
          "scale": Object {
            "range": Array [
              "#54B399",
              "#6092C0",
              "#D36086",
              "#9170B8",
              "#CA8EAE",
              "#D6BF57",
              "#B9A888",
              "#DA8B45",
              "#AA6556",
              "#E7664C",
            ],
          },
          "title": "Rule severity",
        },
        "opacity": Object {
          "condition": Object {
            "selection": "series",
            "value": 1,
          },
          "value": 0.2,
        },
        "tooltip": Array [
          Object {
            "aggregate": "sum",
            "axis": Object {
              "grid": true,
            },
            "field": "finding",
            "title": "Findings",
            "type": "quantitative",
          },
          Object {
            "field": "time",
            "format": "%Y-%m-%d %H:%M:%S",
            "timeUnit": Object {
              "step": 1,
              "unit": "yearmonthdatehoursminutes",
            },
            "title": "Time",
            "type": "temporal",
          },
          Object {
            "field": "",
            "title": "Rule severity",
          },
        ],
        "x": Object {
          "axis": Object {
            "format": "%Y-%m-%d %H:%M",
            "grid": false,
          },
          "field": "time",
          "scale": Object {
            "domain": undefined,
          },
          "timeUnit": Object {
            "step": 1,
            "unit": "yearmonthdatehoursminutes",
          },
          "title": "",
          "type": "temporal",
        },
        "y": Object {
          "aggregate": "sum",
          "axis": Object {
            "grid": true,
          },
          "field": "finding",
          "title": "Count",
          "type": "quantitative",
        },
      },
      "mark": Object {
        "clip": true,
        "type": "bar",
      },
      "selection": Object {
        "series": Object {
          "bind": "legend",
          "encodings": Array [
            "color",
          ],
          "on": "click",
          "type": "multi",
        },
      },
    },
  ],
}
`;

exports[`helper utilities spec tests getOverviewVisualizationSpec function  - snapshot test: should match overview spec 1`] = `
Object {
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "config": Object {
    "legend": Object {
      "labelColor": "#343741",
      "labelFont": "\\"Inter UI\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\"",
      "labelFontSize": 14,
      "rowPadding": 6,
      "titleColor": "#1a1c21",
      "titleFontSize": 14,
      "titleFontWeight": 600,
      "titleLineHeight": 21,
      "titlePadding": 10,
    },
    "view": Object {
      "stroke": "transparent",
    },
  },
  "data": Object {
    "values": Array [],
  },
  "description": "Plot showing average data with raw values in the background.",
  "layer": Array [
    Object {
      "encoding": Object {
        "color": Object {
          "scale": null,
          "value": "#6092C0",
        },
        "tooltip": Array [
          Object {
            "aggregate": "sum",
            "axis": Object {
              "grid": true,
            },
            "field": "finding",
            "title": "Findings",
            "type": "quantitative",
          },
          Object {
            "field": "time",
            "format": "%Y-%m-%d %H:%M:%S",
            "timeUnit": Object {
              "step": 1,
              "unit": "yearmonthdatehoursminutes",
            },
            "title": "Time",
            "type": "temporal",
          },
        ],
        "x": Object {
          "axis": Object {
            "format": "%Y-%m-%d %H:%M",
            "grid": false,
          },
          "field": "time",
          "scale": Object {
            "domain": undefined,
          },
          "timeUnit": Object {
            "step": 1,
            "unit": "yearmonthdatehoursminutes",
          },
          "title": "",
          "type": "temporal",
        },
        "y": Object {
          "aggregate": "sum",
          "axis": Object {
            "grid": true,
          },
          "field": "finding",
          "title": "Count",
          "type": "quantitative",
        },
      },
      "mark": Object {
        "clip": true,
        "type": "bar",
      },
    },
    Object {
      "encoding": Object {
        "tooltip": Array [
          Object {
            "aggregate": "sum",
            "axis": Object {
              "grid": true,
            },
            "field": "alert",
            "title": "Alerts",
            "type": "quantitative",
          },
          Object {
            "field": "time",
            "format": "%Y-%m-%d %H:%M:%S",
            "timeUnit": Object {
              "step": 1,
              "unit": "yearmonthdatehoursminutes",
            },
            "title": "Time",
            "type": "temporal",
          },
        ],
        "x": Object {
          "axis": Object {
            "format": "%Y-%m-%d %H:%M",
            "grid": false,
          },
          "band": 0.5,
          "field": "time",
          "scale": Object {
            "domain": undefined,
          },
          "timeUnit": Object {
            "step": 1,
            "unit": "yearmonthdatehoursminutes",
          },
          "title": "",
          "type": "temporal",
        },
        "y": Object {
          "aggregate": "sum",
          "axis": Object {
            "grid": true,
          },
          "field": "alert",
          "title": "Count",
          "type": "quantitative",
        },
      },
      "mark": Object {
        "clip": true,
        "color": "#ff0000",
        "interpolate": "monotone",
        "point": Object {
          "color": "#ff0000",
          "fill": "white",
          "filled": false,
          "size": 50,
        },
        "type": "line",
      },
    },
  ],
}
`;

exports[`helper utilities spec tests getVisualizationSpec function  - snapshot test: should match visualization spec 1`] = `
Object {
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "config": Object {
    "legend": Object {
      "labelColor": "#343741",
      "labelFont": "\\"Inter UI\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\"",
      "labelFontSize": 14,
      "rowPadding": 6,
      "titleColor": "#1a1c21",
      "titleFontSize": 14,
      "titleFontWeight": 600,
      "titleLineHeight": 21,
      "titlePadding": 10,
    },
    "view": Object {
      "stroke": "transparent",
    },
  },
  "data": Object {
    "values": Array [
      Object {
        "xField": 1,
        "yField": 1,
      },
    ],
  },
  "description": "Visualization",
  "layer": Array [
    Object {
      "encoding": Object {
        "opacity": Object {
          "condition": Object {
            "selection": "series",
            "value": 1,
          },
          "value": 0.2,
        },
      },
      "selection": Object {
        "series": Object {
          "bind": "legend",
          "encodings": Array [
            "color",
          ],
          "on": "click",
          "type": "multi",
        },
      },
      "x": Object {
        "field": "xField",
      },
      "y": Object {
        "field": "yField",
      },
    },
  ],
}
`;
